RavenDB-24485 Allow to get schema validation report

haludi opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-24485

### Additional description
This change introduces two new capabilities:
1. Index schema validation inside indexes.  
2. A background operation to collect schema validation results.  

#### Index schema validation
Two functions were added: **SchemaValid** and **SchemaError**.  
- **SchemaValid** checks only whether a document is valid or not and returns immediately. It does not build an error message.  
- **SchemaError** performs full validation in order to collect all error messages.  

Validation is performed against a single defined schema.  
It applies only to database documents and will throw an error if used on arbitrary objects.  

#### Background operation to collect schema validation results
Schema validation results can be collected through a background operation.  
The operation takes as input both the collection to work on and the schema to validate against.  
The final result can be obtained by fetching the operation once it completes.  

The operation output includes:  
- The number of documents checked.  
- The number of errors found.  
- The collected error messages.  

The operation supports several limits:  
- A time limit for the overall operation.  
- A maximum duration for each read transaction.  
- A cap on the number of collected error messages.  

If the error message cap is reached, the operation continues to validate documents but stops collecting further messages.  
It only continues with counting the number of invalid documents.  

Note: if the operation is split into multiple read transactions, a document may, in principle, be counted more than once.


### Type of change
- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?
- [ ] Low 
- [ ] Moderate 
- [x] High
- [ ] Not relevant

### Backward compatibility
- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?
- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update
- [x] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor
- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team
- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?
- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work
- [x] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [ ] No UI work is needed

haludi committed:
RavenDB-24485 Allow to index schema validity

haludi committed:
RavenDB-24485 Allow to get schema validation status for a collection

haludi commented on src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/SchemaValidationHandlerProcessorForValidate.cs:
```diff
@@ -0,0 +1,115 @@
+            
+            while (DateTime.Now < stopTotalTime)
+            {
+                using (context.OpenReadTransaction())
```
done

haludi commented on src/Raven.Server/Documents/Sharding/Handlers/Processors/SchemaValidation/ShardedSchemaValidationHandlerProcessorForValidate.cs:
```diff
@@ -0,0 +1,31 @@
+                OperationType.ValidateSchemaValidation,
+                $"Schema validation for collection '{Parameters.Collection}' '{RequestHandler.DatabaseName}'",
+                detailedDescription: null,
+                commandFactory: (_, _) => new ValidateSchemaValidationOperation.ValidateSchemaValidationCommand(RequestHandler.ShardExecutor.Conventions, Parameters, operationId),
```
done

haludi commented on src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/SchemaValidationHandlerProcessorForValidate.cs:
```diff
@@ -0,0 +1,115 @@
+                            break;
+                    }
+                    if(enumerator.MoveNext() == false)
+                        break;
```
done

haludi commented on src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:
```diff
@@ -176,7 +178,44 @@ public dynamic TimeSeriesNamesFor(dynamic doc)
+                if (doc is DynamicNullObject)
+                    return doc;
+                
+                ThrowInvalidDocType(doc, nameof(SchemaValid));
```
done

haludi committed:
RavenDB-24485 One schema validation for index, fix operation parameter check,

haludi committed:
RavenDB-24485 Throw when validate 'null' in index

haludi commented:
All fails look like studio work

haludi committed:
RavenDB-24485 `ValidateSchemaValidationOperation` propertiy names

ayende commented on src/Raven.Client/Documents/Operations/SchemaValidation/ValidateSchemaValidationOperation.cs:
```diff
@@ -0,0 +1,119 @@
+    public sealed class Parameters
+    {
+        /// <summary>JSON schema definition text. (Required)</summary>
+        public string Schema { get; set; }
```
This would be the JSON Schema itself? Or a reference to its definition?

ayende replied:
I have an issue with the `text` here.

haludi replied:
The alternative is Blittable, but I don't think it is good for a public API.

ppekrol replied:
Remove `text` from the summary

ayende replied:
I actually meant that _name_ of the property. Should be `ShcemaDefinition` or something like that, I think

haludi replied:
Oh, sorry.
I changed it

ayende commented on src/Raven.Client/Documents/Operations/SchemaValidation/ValidateSchemaValidationOperation.cs:
```diff
@@ -0,0 +1,119 @@
+        /// <summary>Total time limit in minutes. (Optional, default 16, must be > 0 when specified)</summary>
+        public int? MaxDurationInMinutes { get; set; }
+
+        /// <summary>Maximum duration per read batch in seconds. (Optional, default 960, must be > 0 when specified)</summary>
```
Why 960?

ayende replied:
And why do you have both max duration and max batch duration set to the same value?

Why have both in the first plcae?

haludi replied:
The batch time limit is to prevent the read transaction from being open for a long time.
The max duration is to limit the operation.
As I see it, the aim of this operation is to get a full report about the data validation.
The reason to provide the possibility to limit it is only to avoid putting stress on the system.
The main stress comes from long read transactions, so I think the best option is to allow limiting their duration.
I set a default just to make sure there is a limitation in case it’s executed accidentally without specifying one.
I think 15 minutes is a good maximum time to limit the read transaction open.

Do you prefer to remove those and just define "number of docs to scan"?

ayende replied:
1/ Users don't care about max read tx length - shouldn't be exposed to them. You can just create a new read tx every 5 sec or so, no?
2/  And yes - number of docs to scan is better. Use _can_ cancel the operation if they want to, right?

haludi replied:
1. I prefer to give the option to run it with a single read transaction, so the result will not potentially include information about a document twice if it was modified during the operation.  
2. Yes, the user can cancel it. I will change it to the number of documents.

ayende replied:
1/ that is a detail the user just doesn't _want_ to know about.
Use a dictionary and that would deal with this, no?

haludi replied:
For errors, yes.  
But I think we should limit the number of errors a user can get.  
If the collection contains millions of documents, and in the edge case all of them are invalid with many errors, I stop collecting error messages and continue only with counting.  
During counting, a document might be processed twice.  
This is more relevant for large databases where validation can take a long time and many modifications can happen during that period.  
I think it’s convenient to have some control over that, not in terms of read transactions, which the user shouldn’t care about, but in terms of batches.  

But I’ll remove that.  
I’ll keep only the configuration for `errorsCount` and the “number of documents to scan.”

ayende replied:
During counting - do we _care_ if a document that was modified is counted twice?
You are either doing this quickly enough that we won't run into this - or this is big enough that it doesn't matter.
Do we care about transactional guarantees here? Or being 1000% accurate?

haludi replied:
> Or being 1000% accurate? 

I guess we can tolerate some error in the counting.
I will remove that

haludi replied:
done

ayende commented on src/Raven.Client/Documents/Operations/SchemaValidation/StartSchemaValidationOperation.cs:
```diff
@@ -0,0 +1,119 @@
+    /// <summary>
+    /// Parameters for schema validation. Schema and Collection are required; other values are optional.
+    /// </summary>
+    public sealed class Parameters
```
Should we have a way to start from a given etag?
A maximum number of docs to scan?

haludi replied:
https://github.com/ravendb/ravendb/pull/21511#discussion_r2410517566

haludi replied:
I don't think the ETag is required since the idea is to get an understanding of what is not valid, rather than to get it for each document.  
In the end, the user can't automate the fix process using the errors, and going over all the error documents and fixing them manually one by one is not practical.  
Maybe using AI.  
Anyway, there is an option to get all of them by index.  

Of course, I can add ETag, but it will be limited to non-sharded databases unless it becomes a list of ETags.  
I guess if we do that, the result should also include the ETag.

ayende replied:
Let's add this anyway. It is cheap for us to do, and it can open doors for the user.
For sharding, just throw a not supported for this. As you said, that is a combination of weird features.

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/SchemaValidationHandlerProcessorForValidate.cs:
```diff
@@ -0,0 +1,121 @@
+
+            using var errorBuilder = new ErrorBuilder(context);
+            
+            while (DateTime.UtcNow < stopTotalTime)
```
No, please use `Stopwatch` instead here. 
This is subject to clock skew, etc.

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/SchemaValidationHandlerProcessorForValidate.cs:
```diff
@@ -0,0 +1,121 @@
+            var errors = new Dictionary<string, string>();
+            
+            var stopTotalTime = DateTime.UtcNow + maxTime;
+            var progressReportRate = TimeSpan.FromMinutes(1);
```
That is pretty long, no?

haludi replied:
What would be a reasonable time for you? something like 10 seconds?

ayende replied:
Yes, sufficient to keep someone interested and sure that this isn't stuck

haludi replied:
I changed it to 5 seconds

ayende commented on src/Raven.Server/Documents/SchemaValidation/SchemaValidationHelper.cs:
```diff
@@ -220,4 +221,36 @@ private static bool TryGetProperty<T>(BlittableJsonReaderObject schemaDefinition
+    {
+        if (blittable.TryGet(SchemaValidatorConstants.AdditionalProperties, out object additionalProperties) == false 
+            || additionalProperties is true)
+            //If additional properties are allowed, no problematic restriction on metadata is can be.
```
I don't understand this comment. 
For that matter - we should _ignore_ the `@metadata` and its content, no? Is that part of the schema?

haludi replied:
If the user defines a restriction for "additionalProperties", any property that is not explicitly defined in the schema will be validated by "additionalProperties", including the "@metadata" property.
Let's say the user restricts "additionalProperties" to be of type "string", then all the "@metadata" properties will fail validation because they are objects.
I don't see why we should prevent the user from defining a schema on the metadata if he want to.
I think that if he explicitly defines a schema on the metadata, he should know what he is doing.
I wanted to handle implicit definitions the same way "additionalProperties" does.

Do you prefer to disallow schema on the metadata?

ayende replied:
The `@metadata` isn't controlled by the user. It is ours. 
We don't want to have replication rejected because we fail schema validation on the metadata, for example.
Or stuff like that. 

Users don't _know_ what is in the metadata, nor are they likely to consider stuff that we may add in the future.
`@metadata` should be *explicitly* not checked.

haludi replied:
I think if the user wants to add a schema on a custom property in the metadata, there is no need to disallow that.
If we allow to change the metadata it is also possible to want it valid - https://docs.ravendb.net/6.2/client-api/session/how-to/get-and-modify-entity-metadata/#modify-the-metadata
But I will change it.
I am going to reject (throw an exception) on any attempt to define a schema on the metadata.

ayende replied:
Yes, that would work.

haludi replied:
done.
I also extended the schema validation functionality to have an exclude properties functionality, and I used it to avoid validating the @metadata

ayende commented on src/Raven.Server/Documents/SchemaValidation/SchemaValidationHelper.cs:
```diff
@@ -220,4 +221,36 @@ private static bool TryGetProperty<T>(BlittableJsonReaderObject schemaDefinition
+        }
+        else
+        {
+            properties.Modifications = new DynamicJsonValue(properties) { [Client.Constants.Documents.Metadata.Key] = new DynamicJsonValue() };
```
What you you doing this?

haludi replied:
To explicitly define a schema on the metadata and avoid having the "additionalProperties" schema validate it.

ayende replied:
Let's drop this.

haludi replied:
I'm not sure you understood me.  
I implemented the SchemaValidator decoupled from RavenDB, so SchemaValidator doesn't know about the metadata.  

The problem is that the user should not be aware of the metadata.  
But if the user defines a schema that allows only documents with the properties Name and Age and disallows additional properties, that schema will fail for any document because it does contain another property, the metadata.  

I can explicitly add code inside SchemaValidator to handle metadata as we want (like ignoring it), but I prefer to handle that at a higher level.  
Instead, I make sure the metadata is valid for the schema definition.  
This is done only once, right before we use it.  
Again, the other option is to add an explicit condition inside SchemaValidator, and I don't think that's elegant.

haludi replied:
https://github.com/ravendb/ravendb/pull/21511#discussion_r2428840694

ayende commented on src/Raven.Server/Documents/SchemaValidation/SchemaValidationHelper.cs:
```diff
@@ -220,4 +221,36 @@ private static bool TryGetProperty<T>(BlittableJsonReaderObject schemaDefinition
+
+        if (blittable.TryGet(SchemaValidatorConstants.Properties, out BlittableJsonReaderObject properties)
+            && properties.Contains(Client.Constants.Documents.Metadata.Key))
+            //If explicit restriction on metadata is configured we don't verify it in this point.
```
what does this mean? what kind of schema would have it here?

haludi replied:
If the user explicitly defines a schema on the metadata, I don't check whether that definition can be valid for a document in RavenDB.
As I see it, if the user wants to define a schema on the metadata, he is free to do so.
Let's say he wants a custom field in it like doc["@metadata"]["myCustomField"], and he wants to make sure it is a **number**.
But schema validation is very flexible, and he can define anything.
He can also define that all properties of the metadata are **numbers**.
That, of course, will not be valid for any RavenDB document.

ayende replied:
I would rather skip that entirely. They are free to define their own `Metadata`  element, `@metadata` is RavenDB's internal and not their job.

haludi replied:
https://github.com/ravendb/ravendb/pull/21511#discussion_r2413778606

ayende commented on test/SlowTests/Server/Documents/Indexing/SchemaValidationIndexingTests.cs:
```diff
@@ -0,0 +1,188 @@
+            select new 
+            {
+                Id = doc.Id,
+                Error = SchemaError(doc)
```
Why separate `SchemaError` vs. `SchemaValid` ?

ayende replied:
Can I do this in JavaScript indexes as well?

haludi replied:
`SchemaError` collects all actual errors and continues to validate even if the document is found to be invalid, so it collects all the errors.
`SchemaValid` doesn't collect errors at all and returns `false` immediately when an invalidity is found.

If you want to project and get the actual error, you can use `SchemaError`
For only filtering, you can use `SchemaValid` (and avoid the additional cost of collecting the error.

haludi replied:
I didn't implement it for the JavaScript index.
I wasn't sure if we wanted it.
I guess you do want it, right?

ayende replied:
Yes, I expect that it would be pretty simple to add to JS.

About the two methods - in what scenario do I just care about `SchemaValid` and _not_ `SchemaError` ?
Assuming that the cost for valid docs for both of them is the same? And that valid docs are the more common ones.
I would almost always want to _show_ what the problem was, no?

haludi replied:
If you want to do some automation, you might only need to know whether the document is valid or not.  
That can be the case when the same error repeats across many documents.  
But you can handle that with SchemaError or by filtering the documents according to the actual data.

I will remove one of them.  
I will remove SchemaError and change the behavior of SchemaValid to return the errors.

haludi replied:
done

haludi committed:
RavenDB-24485 Use Stopwatch instead of current time.

haludi committed:
RavenDB-24485 Parameters.SchemaDefinition name and comment

haludi committed:
RavenDB-24485 Only one index function and support javascript

src/Raven.Server/Documents/Indexes/Static/JavaScriptIndex.cs

haludi committed:
RavenDB-24485 Exclude @metadata

haludi committed:
RavenDB-24485 Add etag to ValidateSchemaValidationOperation

ppekrol commented on src/Raven.Client/Documents/Operations/SchemaValidation/ValidateSchemaValidationOperation.cs:
```diff
@@ -0,0 +1,118 @@
+/// Optional limits (if omitted server defaults are used): MaxErrorMessages=1024, MaxDocumentsToValidate=unlimited.
+/// You may also provide a starting Etag to continue validation from a point in the collection; by default validation starts from the first document.
+/// </summary>
+public sealed class ValidateSchemaValidationOperation : IMaintenanceOperation<OperationIdResult<StartValidateSchemaValidationOperationResult>>
```
can we have a better name? e.g. StartSchemaValidationOperation?

haludi replied:
What do you think about `ValidateSchemaOperation"?

ppekrol commented on src/Raven.Client/Documents/Operations/SchemaValidation/ValidateSchemaValidationOperation.cs:
```diff
@@ -0,0 +1,118 @@
+        public long? MaxDocumentsToValidate { get; set; }
+        
+        /// <summary>Starting document etag to begin validation from. (Optional, default: start from the first document)</summary>
+        public long? Etag { get; set; }
```
StartEtag?

haludi replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/SchemaValidation/ValidateSchemaValidationOperation.cs:
```diff
@@ -0,0 +1,118 @@
+        return new ValidateSchemaValidationCommand(conventions, _parameters);
+    }
+
+    internal class ValidateSchemaValidationCommand : RavenCommand<OperationIdResult<StartValidateSchemaValidationOperationResult>>, IRaftCommand
```
StartSchemaValidationOperation

haludi replied:
"ValidateSchemaCommand"?

ppekrol commented on src/Raven.Client/Documents/Operations/SchemaValidation/ValidateSchemaValidationOperation.cs:
```diff
@@ -0,0 +1,118 @@
+            var operationIdResult = JsonDeserializationClient.OperationIdResult(response);
+
+            // OperationNodeTag used to fetch operation status
+            operationIdResult.OperationNodeTag ??= result.ResponsibleNode;
```
is there a chance that this can be null? I do not think so?

haludi replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/SchemaValidation/ValidateSchemaValidationResult.cs:
```diff
@@ -0,0 +1,79 @@
+    }
+}
+
+public sealed class ValidateSchemaValidationResult : ValidateSchemaValidationProgress, IOperationResult
```
I really do not like that ValidateValidateValidate

haludi replied:
Does "ValidateSchemaResult" sound good?

ppekrol commented on src/Raven.Server/Documents/Indexes/Static/JavaScriptIndex.cs:
```diff
@@ -65,6 +65,7 @@ protected override void OnInitializeEngine(Engine engine)
             engine.SetValue("counterNamesFor", new ClrFunction(_engine, "counterNamesFor", CounterNamesFor));
             engine.SetValue("loadAttachment", new ClrFunction(engine, "loadAttachment", LoadAttachment));
             engine.SetValue("loadAttachments", new ClrFunction(engine, "loadAttachment", LoadAttachments));
+            engine.SetValue("schemaValidate", new ClrFunction(engine, "schemaValidate", SchemaValidate));
```
we are trying to move away from having direct methods here. Can we introduce an object `schema`. Look here:

```
        ObjectInstance embeddingsObject = new JsObject(DocumentScript.ScriptEngine);
        embeddingsObject.FastSetProperty("generate", new PropertyDescriptor(new ClrFunction(DocumentScript.ScriptEngine, "generate", EmbeddingsGenerate), false, false, false));
        DocumentScript.ScriptEngine.SetValue("embeddings", embeddingsObject);
```

haludi replied:
done

haludi committed:
RavenDB-24485 Cleaning code

haludi commented on src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:
```diff
@@ -176,6 +178,30 @@ public dynamic TimeSeriesNamesFor(dynamic doc)
+        {
+            var json = EnsureSourceDocumentOrThrow(doc, nameof(SchemaValid));
+            return CurrentIndexingScope.Current.SchemaValid(json);
+        }
```
@ppekrol, what do you think about that?
Would `SchemaError` be more suitable here?

ppekrol replied:
handle this as we do with NoTracking in the indexes.

My expectation in the definition would be to see:
```
from doc in docs
select {
    IsSchemaValid = Schema.IsValid(doc)
}
```

haludi commented:
@ppekrol, I think for `AbstractSchemaValidationHandlerProcessorForValidate` it is ok to leave it like that.

haludi committed:
RavenDB-24485 Avoid VlidateValidate in names

haludi committed:
RavenDB-24485 Concat `excludedProperties`

haludi committed:
RavenDB-24485 Move javascript schemaValidate func to schema object

	src/Raven.Server/Documents/Indexes/Static/JavaScriptIndex.cs

haludi committed:
RavenDB-24485 Explicitly convert string to ReadOnlySpan<char>

ppekrol commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -287,6 +288,7 @@ public bool DeployedOnAllNodes
         public bool IsOnBeforeExecuteIndexing { get; private set; }
 
         public TestIndexRun TestRun;
+        internal (SchemaValidator validator, IDisposable returnCtx) _schemaValidator;
```
uppercase please :)

haludi replied:
I changed it to a dedicated class

ppekrol commented on src/Raven.Server/Documents/Indexes/Static/JavaScriptIndex.cs:
```diff
@@ -66,6 +66,10 @@ protected override void OnInitializeEngine(Engine engine)
             engine.SetValue("id", new ClrFunction(_engine, "id", GetDocumentId));
+            
+            var schemaValidatorObject = new JsObject(engine);
+            schemaValidatorObject.FastSetProperty("validate", new PropertyDescriptor(new ClrFunction(engine, "validate", SchemaValidate), false, false, false));
```
rebase and use `SetClrFunc` instead

haludi replied:
I did not find an existing implementation for an object, so I implemented it myself.
I did not change other usages to use my implementation so as not to dirty my PR too much.

ppekrol commented on src/Raven.Server/Documents/Operations/OperationType.cs:
```diff
@@ -80,5 +80,8 @@ public enum OperationType
+    AdoptOrphanedRevisions,
+    
+    [Description("Validate Schema Validation")]
+    ValidateSchemaValidation
```
just ValidateSchema ?

haludi replied:
done

ppekrol commented on src/Raven.Client/Documents/Indexes/IndexDefinition.cs:
```diff
@@ -98,6 +98,8 @@ public virtual IndexSourceType SourceType
 
         public virtual ArchivedDataProcessingBehavior? ArchivedDataProcessingBehavior { get; set; }
+        
+        public string SchemaValidation { get; set; }
```
internal?

Also shouldn't it be SchemaValidationConfiguration, exactly what we have in the DatabaseRecord? So we can handle multiple collections?

haludi replied:
Why "internal"?

`SchemaValidationConfiguration` has `Disabled` properties that are not relevant.
I think it is better to have a dedicated class for `IndexDefinition`
I defined `IndexSchemaDefinitions` class for that

haludi committed:
RavenDB-24485 Avoid "Validate..Validate" in names

	modified:   src/Raven.Client/Documents/Operations/SchemaValidation/ValidateSchemaResult.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/AbstractSchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/SchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Documents/Handlers/SchemaValidationHandler.cs
	modified:   src/Raven.Server/Documents/Indexes/Index.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/CurrentIndexingScope.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/JavaScriptIndex.cs
	modified:   src/Raven.Server/Documents/Operations/OperationType.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaValidatorCache.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/Validators/Object/ObjectSchemaRuleValidator.cs
	modified:   src/Raven.Server/Documents/Sharding/Handlers/Processors/SchemaValidation/ShardedSchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Documents/Sharding/Handlers/ShardedSchemaValidationHandler.cs
	modified:   src/Raven.Server/Json/JsonDeserializationServer.cs
	modified:   test/SlowTests/SchemaValidation/SchemaValidationOperationTests.cs

haludi committed:
RavenDB-24485 Index array of errors and index configuration per collection

	modified:   src/Raven.Client/Documents/Indexes/IndexDefinition.cs
	new file:   src/Raven.Client/Documents/Indexes/IndexSchemaDefinitions.cs
	modified:   src/Raven.Server/Documents/Indexes/Index.cs
	modified:   src/Raven.Server/Documents/Indexes/IndexDefinitionBaseServerSide.cs
	new file:   src/Raven.Server/Documents/Indexes/SchemaValidation/IndexSchemaValidatorsByCollection.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/CurrentIndexingScope.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/JavaScriptIndex.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/MapIndexDefinition.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/Roslyn/MapFunctionProcessor.cs
	new file:   src/Raven.Server/Documents/Indexes/Static/Roslyn/Rewriters/SchemaValidatorRewriter.cs
	modified:   src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs
	modified:   src/Raven.Server/Documents/Patch/JavaScriptUtils.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ErrorMessage/AbstractBuffer.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ErrorMessage/ErrorBuffer.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ErrorMessage/ErrorBuilder.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ErrorMessage/IErrorBuffer.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/SchemaValidatorCache.cs
	modified:   src/Raven.Server/Documents/SchemaValidation/ValidationPath.cs
	modified:   src/Raven.Server/Extensions/JsonSerializationExtensions.cs
	modified:   src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs
	modified:   src/Raven.Server/Json/Sync/BlittableJsonTextWriterExtensions.cs
	modified:   src/Raven.Server/ServerWide/ServerStore.cs
	modified:   test/FastTests/SchemaValidation/SchemaValidationTestsBase.cs
	modified:   test/SlowTests.Issues/Issues/RavenDB-17597.cs
	modified:   test/SlowTests/SchemaValidation/BasicIntegrationSchemaValidationTests.cs
	modified:   test/SlowTests/SchemaValidation/SchemaValidationOperationTests.cs
	modified:   test/SlowTests/Server/Documents/Indexing/SchemaValidationIndexingTests.cs

haludi commented:
I changed the index to index an array of errors.
Also, the index schema definition is now a schema per collection.
If no schema is defined for the index, the database schema is used.
If no schema is defined in the database, that means any content is valid.

* If the index relies on the database schema and it is changed, we don't index already-indexed documents.
* The user will need to reset it manually

haludi commented:
The only failing test now is related to the studio
FastTests.Issues.RavenDB_6250.All_operations_has_details_providers

haludi committed:
RavenDB-24485 Rename ValidateSchemaOperation to StartSchemaValidationOperation and update related references

	renamed:    src/Raven.Client/Documents/Operations/SchemaValidation/ValidateSchemaOperation.cs -> src/Raven.Client/Documents/Operations/SchemaValidation/StartSchemaValidationOperation.cs
	modified:   src/Raven.Server/Documents/Handlers/Processors/SchemaValidation/AbstractSchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Documents/Sharding/Handlers/Processors/SchemaValidation/ShardedSchemaValidationHandlerProcessorForValidate.cs
	modified:   src/Raven.Server/Json/JsonDeserializationServer.cs
	modified:   test/SlowTests/SchemaValidation/SchemaValidationOperationTests.cs

ppekrol commented on test/SlowTests/Server/Documents/Indexing/SchemaValidationIndexingTests.cs:
```diff
@@ -0,0 +1,425 @@
+        select new 
+        {
+            Id = doc.Id,
+            Errors = Schema.GetErrorsFor(doc)
```
We also need this in the AbstractIndexCreationTask, right?

haludi replied:
As I understand, we didn't want to add that.
But of course I can do that.

ppekrol commented on src/Raven.Server/Documents/Operations/OperationType.cs:
```diff
@@ -80,5 +80,8 @@ public enum OperationType
-    AdoptOrphanedRevisions
+    AdoptOrphanedRevisions,
+    
+    [Description("Validate Schema Validation")]
```
Validate JSON Schema ?

haludi replied:
done

ppekrol reviewed (commented):
Let's fix last commit, it violates commit conventions

haludi committed:
RavenDB-24485 Disallow revert revision when shema validation is enabled.

haludi commented:
I added to this PR a check to disallow a revert revision when the schema is enabled.
This is affecting both revert by ids and revert by time.
* Revert by time includes the collections in the request parameter so it will be rejected only if the relevant collections are enabled.
* For revert by ids it will always reject.

We consider revert revision as a maintenance operation so it makes sense to require explicitly disabling the schema for that.

M4xymm committed:
RavenDB-25473 working notification for validate schema with grid

RavenDB-25473 (WIP) - feat (schema validation): Added validation schema view sheet panel with progress tracking and test settings configuration. Updated related operations and UI enhancements for schema validation.

RavenDB-25473 feat (document schema playground): Introduced improved schema validation with support for multiple validators, dynamic progress tracking, and enhanced UI components. Refactored unused code and added collapsible panels.

RavenDB-25473 feat (document schema validation): Refactored validation operations to enhance test configuration, progress monitoring, and error presentation. Introduced reusable components for validation status and user interaction. Added operation abort functionality and improved table column rendering.

M4xymm committed:
RavenDB-25473 feat (schema validation): Enhanced progress tracking for schema validation operations with improved state handling (`isCompleted`, `isErrored`, `isLoading`). Refactored UI for test settings and validation monitoring.

M4xymm commented:
UI Work: 

<img width="1862" height="1230" alt="image" src="https://github.com/user-attachments/assets/ad8ca748-fe42-4bed-862a-f9233182edf6" />

<img width="1882" height="1225" alt="image" src="https://github.com/user-attachments/assets/3795029f-1fb3-4e83-a38d-077d42cc38c5" />

<img width="1899" height="1271" alt="image" src="https://github.com/user-attachments/assets/29476ae4-6ea3-4783-a9cd-fe660f932004" />

<img width="1056" height="1158" alt="image" src="https://github.com/user-attachments/assets/ddbbc540-2708-437f-bec1-4c7ff9e4233a" />

<img width="1853" height="1223" alt="image" src="https://github.com/user-attachments/assets/f06c1adc-e100-4a28-a7ed-ddfdeed875ac" />

<img width="1884" height="1226" alt="image" src="https://github.com/user-attachments/assets/69bfd430-4a78-4c46-8d89-ece36ded4053" />

ppekrol commented:
@haludi can you rebase and fix failing tests?

kalczur commented on src/Raven.Studio/typescript/commands/database/settings/validateSchemaCommand.ts:
```diff
@@ -0,0 +1,20 @@
+import endpoints = require("endpoints");
+
+class validateSchemaCommand extends commandBase {
+    constructor(private db: database | string, private validateSchemaRequestDto: ValidateSchemaRequestDto) {
```
```suggestion
    constructor(private db: database | string, private dto: ValidateSchemaRequestDto) {
```

kalczur commented on src/Raven.Studio/typescript/components/pages/database/settings/documentSchema/partials/DocumentSchemaPlayground.tsx:
```diff
@@ -14,24 +14,61 @@ import {
-                    <Col>
+            <Row className="gy-sm">
+                <Col>
+                    <form>
```
add onSubmit

kalczur commented on src/Raven.Studio/typescript/components/pages/database/settings/documentSchema/partials/DocumentSchemaPlayground.tsx:
```diff
@@ -43,7 +80,7 @@ export default function DocumentSchemaPlayground() {
                                 Cancel
                             </a>
-                            <Button className="rounded-pill">
+                            <Button type="submit" onClick={handleSubmit(handleOpenSheet)} className="rounded-pill">
```
remove onClick

kalczur commented on src/Raven.Studio/typescript/components/pages/database/settings/documentSchema/partials/DocumentSchemaSelectActions.tsx:
```diff
@@ -42,8 +46,8 @@ export default function DocumentSchemaSelectActions() {
     } = useBoolean(false);
-    const [operationConfirm, setOperationConfirm] = React.useState<OperationConfirm>(null);
     const confirm = useConfirm();
+    const [operationConfirm, setOperationConfirm] = React.useState<OperationConfirm>(null);
```
We alawys omit `React.` part in this case
```suggestion
    const [operationConfirm, setOperationConfirm] = useState<OperationConfirm>(null);
```

kalczur commented on src/Raven.Studio/typescript/components/pages/database/settings/documentSchema/partials/ValidationSchemaViewSheetPanel.tsx:
```diff
@@ -0,0 +1,583 @@
+    const collections = useAppSelector(collectionsTrackerSelectors.collections);
+
+    const form = useForm<ValidateSchemaFormData>({
+        resolver: yupResolver(schema),
```
missing defaultValues

M4xymm replied:
In this case, default values are not necessary

kalczur commented on src/Raven.Studio/typescript/components/pages/database/settings/documentSchema/partials/ValidationSchemaViewSheetPanel.tsx:
```diff
@@ -0,0 +1,583 @@
+interface ValidationOperationProgress extends ValidateSchemaResult {
+    isCompleted?: boolean;
+    isErrored?: boolean;
+    isLoading?: boolean;
```
`state: "comple" | "error" | "loading"`

kalczur commented on src/Raven.Studio/typescript/components/pages/database/settings/documentSchema/partials/ValidationSchemaViewSheetPanel.tsx:
```diff
@@ -0,0 +1,583 @@
+            documentSchemaUtils.mapToValidateSchemaRequestDto(validator, formData)
+        );
+
+        return await Promise.all(dtos.map(async (dto) => await databasesService.validateSchema(databaseName, dto)));
```
maybe `Promise.allSettled()`?

kalczur commented on src/Raven.Studio/typescript/components/pages/database/settings/documentSchema/partials/ValidationSchemaViewSheetPanel.tsx:
```diff
@@ -0,0 +1,583 @@
+        control,
+    });
+
+    const isValidating = useMemo(() => {
```
validators is an array so it will be recalculated on every rerender
you can omit useMemo

kalczur commented on src/Raven.Studio/typescript/components/pages/database/settings/documentSchema/partials/ValidationSchemaViewSheetPanel.tsx:
```diff
@@ -0,0 +1,583 @@
+        getSortedRowModel: getSortedRowModel(),
+        getFilteredRowModel: getFilteredRowModel(),
+    });
+    return <VirtualTable className="panel-bg-1" isLoading={loading} table={table} heightInPx={400} />;
```
calculate height based on data

kalczur commented on src/Raven.Studio/typescript/components/pages/database/settings/documentSchema/partials/ValidationSchemaViewSheetPanel.tsx:
```diff
@@ -0,0 +1,583 @@
+        ];
+
+        return cols;
+    }, []);
```
Resize will work in this case? Maybe wrap getSize in useCallback and pass it in dependency array

M4xymm committed:
RavenDB-25473 feat (schema validation): Refactored validation command and UI components. Improved error handling, state management, and introduced `Promise.allSettled` for async operations. Updated `DocumentSchemaPlayground` to better support form submission.

M4xymm committed:
RavenDB-25473 feat (schema validation): Replaced separate state flags (`isCompleted`, `isErrored`, `isLoading`) with a unified `status` field.

M4xymm committed:
RavenDB-24485 feat (operation types): Added `ValidateSchema` to the list of supported operation types.

M4xymm committed:
RavenDB-25473 feat (schema validation): Introduced dynamic table height calculation in `ValidationSchemaViewSheetPanel` using `virtualTableConstants`. Simplified column definition logic.

ppekrol commented:
@M4xymm can you rebase and resolve conflict?

M4xymm committed:
RavenDB-25473 fix (document schema playground): Temporarily disable `DocumentSchemaPlaygroundAboutView` component rendering.

M4xymm commented:
> @M4xymm can you rebase and resolve conflict?

Sure :), Done. Additionally, I disabled the “About view” rendering until it is done. [RDoc-3604](https://issues.hibernatingrhinos.com/issue/RDoc-3604)
